Optimizations for large builds

aras-p opened the pull request:
Redo how the data is gathered & stored to better scale for large builds (ref #10, #30).

TLDR: **3x faster, 15x lower memory usage, 4x smaller file size**.

Previously, `--stop` was just smashing all input JSON files together, and `--analyze` was parsing that huge JSON file, and performing the analysis. The memory allocator used was a simple "bump a pointer, never free" allocator. On a decent size project (e.g. Unity editor build), that was taking 4.8sec / 5.8GB for stop, and 9.9sec / 7.3GB for analyze (so **total 14.7sec**, and max **7.3GB memory** usage). The resulting file was **1.01GB size**.

Now:

- [x] do the JSON parsing during `--stop` step into the build events data structures (which includes string/name deduplication), and as a result just store the binary data structure file. `--analyze` reads that binary file and does no JSON parsing at all now.
- [x] use regular built-in memory allocator (saves a ton of memory for large analyses).
- [x] multi-threaded `--stop` part that parses all the JSON files, using [enkiTS](https://github.com/dougbinks/enkiTS) task scheduler.
- [x] replace `sajson` JSON parser with [simdjson](https://github.com/lemire/simdjson).
- [x] use [bytell_hash_map](https://probablydance.com/2018/05/28/a-new-fast-hash-table-in-response-to-googles-new-fast-hash-table/) instead of C++ built-in one.
- [x] use [xxHash](https://github.com/Cyan4973/xxHash) hash function instead of C++ built-in one.
- [x] optimize "Template sets" and "Function sets" analysis calculation

On the same Unity editor build: 2.4sec / 0.35GB for stop, and 2.7sec / 0.51GB for analyze (so **total 5.1sec**, and max **0.5GB memory** usage). The resulting file is **280MB size**.

However the data files produced by `--stop` and used by `--analyze` are no longer compatible with previous versions (previous were just JSON, now custom binary).

aras-p committed:
Print memory allocated.

On a snapshot of editor build:
--stop: 4.8s (5834.4MB alloc),
--analyze: 9.9s (7261.7MB alloc).

aras-p committed:
Optimize: --stop produces binary data dump instead of JSON; JSON parsing and name deduplication happens during --stop, not during --analyze.
On a snapshot of unity editor build:
--stop: 4.8s 5834MB -> 9.5s 7260MB
--analyze: 9.9s 7262MB -> 4.5s 1270MB
data file: 1.01GB -> 268MB

aras-p committed:
Encapsulate internal JSON parsing & data building state, saves a bit of memory too:
On a snapshot of unity editor build:
--stop: 9.5s 7260MB -> 9.5s 6968MB
--analyze: 4.5s 1270MB (unchanged)
data file: 268MB (unchanged)

aras-p committed:
Use precomputed string hashes in nameToIndex map:
On a snapshot of unity editor build:
--stop: 9.5s 6968MB -> 9.3s 6975MB

aras-p committed:
Replace unordered_map for name deduplication with Malte Skarupke's bytell_hash_map (https://probablydance.com/2018/05/28/a-new-fast-hash-table-in-response-to-googles-new-fast-hash-table/)
On a snapshot of unity editor build:
--stop: 9.3s 6975MB -> 8.2s 6984MB

aras-p committed:
Reuse string buffer memory for reading JSON files into.
On a snapshot of unity editor build:
--stop: 8.2s 6984MB -> 7.8s 6071MB

aras-p committed:
Optimize file writing by using a buffer instead of tiny fwrite calls.
On a snapshot of unity editor build:
--stop: 7.8s 6071MB -> 6.5s 6071MB

aras-p committed:
Optimize file reading by not doing tiny fread calls.
On a snapshot of unity editor build:
--analyze: 4.5s 1270MB -> 3.4s 1519MB

aras-p committed:
Optimize EmitCollapsedTemplateOpt.
On a snapshot of unity editor build:
--analyze: 3.4s 1519MB -> 3.3s 1437MB

aras-p committed:
Fixing Linux & Windows builds

ben-craig commented on src/Analysis.cpp:
```diff
@@ -360,9 +361,17 @@ void Analysis::EmitCollapsedTemplates()
+    
     for (const auto& fn : functions)
     {
-        auto &stats = collapsed[collapseName(llvm::demangle(GetBuildName(fn.first.first)))];
```
I guess this was showing up in profiles?

When writing this, it did cross my mind that I was doing potentially redundant demangling.  I didn't cache the demangling at the time though, because I was worried about the memory consumption.  If that looks fine on Unity, then it looks fine to me.

aras-p replied:
It was showing up a bit indeed (not much). I've checked that the produced results are the same in all tests & in the full unity build analysis, so Should Be Fine :)

ben-craig commented on src/Analysis.cpp:
```diff
@@ -449,7 +458,7 @@ void Analysis::EndAnalysis()
                 dname = dname.substr(0, config.maxName-2) + "...";
             int ms = int(e.second.us / 1000);
-            int avg = int(ms / e.second.count);
+            int avg = int(ms / std::max(e.second.count,1));
```
I guess I just never ran into the div zero case... good catch.

ben-craig reviewed (commented):
aras-p doesn't optimize runtime code.
aras-p doesn't optimize build throughput.
aras-p optimizes the build throughput profiler.

I'm now expecting performance patches for whatever perf visualizer you use.

aras-p committed:
Remove "bump pointer, never free" allocator. It's fast, but huge memory usage on large codebases is a problem, and is not thread safe either.
On a snapshot of unity editor build:
--stop: 6.5s 6071MB -> 7.0s 493MB
--analyze: 3.3s 1437MB -> 4.7s 779MB

aras-p committed:
Add enkiTS task scheduler library from master branch 3686b99

aras-p committed:
Add xxHash hashing library from dev branch 5dd837e

aras-p committed:
Use arena allocator for nameToIndex string names, and xxHash (XXH64) hash function.
On a snapshot of unity editor build:
--stop: 7.0s 493MB -> 5.4s 466MB; BuildEventsParser::NameToIndex 1.47s -> 960ms

aras-p committed:
Use arena allocator for all BuildNames and other names produced during analysis.
On a snapshot of unity editor build:
--stop: 5.4s 466MB -> 5.4s 342MB
--analyze: 4.7s 779MB -> 4.1s 750MB

aras-p committed:
Replace sajson with simdjson (master branch 140e4dd)

aras-p committed:
Change code to use simdjson
On a snapshot of unity editor build:
--stop: 5.4s 342MB -> 5.1s 477MB

aras-p committed:
Multi-threaded JSON parsing
On a snapshot of unity editor build (MacBookPro 2018, 6c/12t):
--stop: 5.1s -> 2.4s

aras-p committed:
Update Make & CMake projects

aras-p committed:
Set timeout for Github Actions (5min), and test CMake path in Linux builder too

aras-p committed:
Fixing cmake test action

aras-p committed:
Actions: try to use gcc 7 on Linux

aras-p committed:
Actions: try to use gcc 7 on Linux

aras-p committed:
Actions: try to use gcc 7 on Linux

aras-p committed:
Windows: switch to VS2019 and fix project files

aras-p committed:
Update readme

aras-p committed:
Linux: link to -lpthread

aras-p committed:
Fix double fclose

aras-p committed:
Make data file have header "magic" and a checksum hash, to avoid blindly loading wrong files

aras-p committed:
Move filename prettifying to --stop phase instead of doing it in --analyze
On a snapshot of unity editor build:
--stop: 2.4s -> 2.3s, 368MB
--analyze: 4.4s 750MB -> 4.4s, 673MB

aras-p committed:
Move name demangling to --stop phase instead of doing it in --analyze
On a snapshot of unity editor build:
--stop: 2.3s 368MB -> 2.4s, 348MB
--analyze: 4.4s 673MB -> 3.6s 641MB

aras-p committed:
Use ska::bytell_hash_map in analysis phase as well.
On a snapshot of unity editor build:
--stop: unchanged at 2.4s 348MB
--analyze: 3.6s 641MB -> 2.9s 578MB

aras-p committed:
In "Function sets that took longest to compile / optimize" (EmitCollapsedTemplateOpt), only actually process templated functions.
This makes it faster (skips over many more functions without attempting to collapse the template name), and also
"fixes" output where function sets are actual sets and not random slow individual functions.

On a snapshot of unity editor build:
--stop: unchanged at 2.4s 348MB
--analyze: 2.9s 578MB -> 2.7s 505MB

aras-p committed:
Fixing Linux build
